server/src/main/java/org/elasticsearch/repositories/blobstore/ChunkedBlobOutputStream.java

```diff
@@ -0,0 +1,149 @@
+import org.elasticsearch.common.util.BigArrays;
+import org.elasticsearch.core.Releasables;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Base class for doing chunked writes to a blob store. Some blob stores require either up-front knowledge of the size of the blob that
+ * will be written or writing it in chunks that are then joined into the final blob at the end of the write. This class provides a basis
+ * on which to implement an output stream that encapsulates such a chunked write.
+ *
+ * @param <T> type of chunk identifier
+ */
+public abstract class ChunkedBlobOutputStream<T> extends OutputStream {
+
+    /**
+     * List of identifiers of already written chunks.
+     */
+    protected final List<T> parts = new ArrayList<>();
+
+    /**
+     * Size of the write buffer above which it must be flushed to storage.
+     */
+    private final long maxBytesToBuffer;
+
+    /**
+     * Big arrays to be able to allocate buffers from pooled bytes.
+     */
+    private final BigArrays bigArrays;
+
+    /**
+     * Current write buffer.
+     */
+    protected ReleasableBytesStreamOutput buffer;
+
+    /**
+     * Set to true once no more calls to {@link #write} are expected and the blob has been received by {@link #write} in full so that
+     * {@link #close()} knows whether to clean up existing chunks or finish a chunked write.
+     */
+    protected boolean successful = false;
+
+    /**
+     * Is set to {@code true} once this stream has been closed.
+     */
+    private boolean closed = false;
+
+    /**
+     * Number of bytes flushed to blob storage so far.
+     */
+    protected long flushedBytes = 0L;
+
+    protected ChunkedBlobOutputStream(BigArrays bigArrays, long maxBytesToBuffer) {
+        this.bigArrays = bigArrays;
+        this.maxBytesToBuffer = maxBytesToBuffer;
+        buffer = new ReleasableBytesStreamOutput(bigArrays);
+    }
+
+    @Override
+    public final void write(int b) throws IOException {
+        buffer.write(b);
+        maybeFlushBuffer();
+    }
+
+    @Override
+    public final void write(byte[] b, int off, int len) throws IOException {
+        buffer.write(b, off, len);
+        maybeFlushBuffer();
+    }
+
+    @Override
+    public final void close() throws IOException {
+        if (closed) {
+            assert false : "this output stream should only be closed once";
+            throw new AlreadyClosedException("already closed");
+        }
+        closed = true;
+        try {
+            if (successful) {
+                onCompletion();
+            } else {
+                onFailure();
+            }
+        } finally {
+            Releasables.close(buffer);
+        }
+    }
+
+    /**
+     * Mark all blob bytes as properly received by {@link #write}, indicating that {@link #close} may finalize the blob.
+     */
+    public final void markSuccess() {
```

Comment:
Really a suggestion:
```suggestion
    public final void done() {
```